src/psyclone/transformations.py

```diff
@@ -563,6 +566,394 @@ def apply(self, node, options=None):
+                  transformation.
+        :rtype: (:py:class:`psyclone.psyir.nodes.Schedule, \
+                 :py:class:`psyclone.undoredo.Memento`)
+        '''
+
+        self.validate(node, options=options)
+
+        schedule = node.root
+        # create a memento of the schedule and the proposed transformation
+        keep = Memento(schedule, self, [node])
+
+        # Find all the OpenMP Single & Master regions
+        task_regions = node.walk(OMPSerialDirective)
+
+        # Loop over the task regions
+        for task_region in task_regions:
+            create_endtaskwait = False
+            endwaits = []
+            # Find all of the taskloops
+            taskloops = task_region.walk(OMPTaskloopDirective)
+            # Get the positions of all of the taskloops
+            taskloop_positions = [-1] * len(taskloops)
+            # Get the forward_dependence position of all of the taskloops
+            dependence_position = [None] * len(taskloops)
+            for i, taskloop in enumerate(taskloops):
+                taskloop_positions[i] = taskloop.abs_position
+                # Only set forward_dep for taskloops with nogroup set
+                forward_dep = None
+                if taskloop.nogroup:
+                    forward_dep = OMPTaskwaitTrans.get_forward_dependence(
+                            taskloop, node)
+                    # If the forward_dependence is one of our parents then we
+                    # should ignore it
+                    if (forward_dep is not None and
+                            forward_dep.abs_position < taskloop.abs_position):
+                        # If we're in a blocking single region and the
+                        # dependency for any of its tasks points to its
+                        # parent single region, then its "real" next dependency
+                        # is outside of the single region. To ensure we
+                        # synchronize before this dependency, we must have a
+                        # task synchronization construct before the spawning
+                        # thread leaves the single region. It is possible that
+                        # this dependency could be handled by another
+                        # intermediary taskwait, however I can't work out a
+                        # good way to detect that now (since there is no
+                        # "end of single region" abs_position to map to).
+                        # I could store a list of all the taskloops that
+                        # require this final taskwait, and walk from each one
+                        # to see if a taskwait has been placed to satisfy their
+                        # dependency, but that does not seem elegant somehow.
+                        if forward_dep is task_region:
+                            endwaits.append(taskloop)
+                            create_endtaskwait = True
+                        forward_dep = None
```

Comment:
Could just do `continue` here to make it a bit more obvious?

Reply:
Fixed